File: engine/src/main/java/de/hanno/hpengine/engine/graphics/shader/Shader.java
package de.hanno.hpengine.engine.graphics.shader;

import de.hanno.hpengine.engine.DirectoryManager;
import de.hanno.hpengine.engine.Engine;
import de.hanno.hpengine.engine.graphics.renderer.GpuContext;
import de.hanno.hpengine.engine.graphics.shader.define.Defines;
import de.hanno.hpengine.util.ressources.CodeSource;
import org.apache.commons.io.FileUtils;
import org.lwjgl.opengl.*;
import de.hanno.hpengine.util.TypedTuple;
import de.hanno.hpengine.util.Util;
import de.hanno.hpengine.util.ressources.Reloadable;

import java.io.File;
import java.io.IOException;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public interface Shader extends Reloadable {

    Logger LOGGER = Logger.getLogger(Shader.class.getName());

    static <SHADERTYPE extends Shader> SHADERTYPE loadShader(GpuContext gpuContext, Class<SHADERTYPE> type, String filename) throws Exception {
        return loadShader(gpuContext, type, new File(getDirectory() + filename), new Defines());
    }

    static <SHADERTYPE extends Shader> SHADERTYPE loadShader(GpuContext gpuContext, Class<SHADERTYPE> type, File file) throws Exception {
        return loadShader(gpuContext, type, file, new Defines());
    }

    static <SHADERTYPE extends Shader> SHADERTYPE loadShader(GpuContext gpuContext, Class<SHADERTYPE> type, File file, Defines defines) throws Exception {
        return loadShader(gpuContext, type, new CodeSource(file), defines);
    }

    static <SHADERTYPE extends Shader> SHADERTYPE loadShader(GpuContext gpuContext, Class<SHADERTYPE> type, CodeSource shaderSource) {
        return loadShader(gpuContext, type, shaderSource, new Defines());
    }

    static <SHADERTYPE extends Shader> SHADERTYPE loadShader(GpuContext gpuContext, Class<SHADERTYPE> type, CodeSource shaderSource, Defines defines) {

        if (shaderSource == null) {
            LOGGER.severe("Shadersource null, returning null de.hanno.hpengine.shader");
            return null;
        }

        String resultingShaderSource = "#version 430 core \n"
                + (defines.isEmpty() ? "" : defines) + "\n"
                + ShaderDefine.getGlobalDefinesString() + "\n";

        String findStr = "\n";
        int newlineCount = (resultingShaderSource.split(findStr, -1).length - 1);

        String actualShaderSource = shaderSource.getSource();

        try {
            TypedTuple<String, Integer> tuple = replaceIncludes(actualShaderSource, newlineCount);
            actualShaderSource = tuple.getLeft();
            newlineCount = tuple.getRight();
            resultingShaderSource += actualShaderSource;
        } catch (IOException e) {
            e.printStackTrace();
        }

        SHADERTYPE shader = null;
        try {
            shader = type.newInstance();
            shader.setShaderSource(shaderSource);
        } catch (Exception e) {
            e.printStackTrace();
        }

        final SHADERTYPE finalShader = shader;
        final int[] shaderID = new int[1];
        final SHADERTYPE finalShader1 = shader;
        final String finalResultingShaderSource = resultingShaderSource;
        gpuContext.execute(() -> {
            shaderID[0] = GL20.glCreateShader(finalShader.getShaderType().glShaderType);
            finalShader1.setId(shaderID[0]);
            GL20.glShaderSource(shaderID[0], finalResultingShaderSource);
            GL20.glCompileShader(shaderID[0]);
        });

        shaderSource.setResultingShaderSource(resultingShaderSource);

        final boolean[] shaderLoadFailed = new boolean[1];
        final int finalNewlineCount = newlineCount;
        gpuContext.execute(() -> {
            if (GL20.glGetShaderi(shaderID[0], GL20.GL_COMPILE_STATUS) == GL11.GL_FALSE) {
                System.err.println("Could not compile " + type.getSimpleName() + ": " + shaderSource.getFilename());
                String shaderInfoLog = GL20.glGetShaderInfoLog(shaderID[0], 10000);
                shaderInfoLog = replaceLineNumbersWithDynamicLinesAdded(shaderInfoLog, finalNewlineCount);
                System.err.println(shaderInfoLog);
                shaderLoadFailed[0] = true;
            }
        });

        if (shaderLoadFailed[0]) {
            throw new ShaderLoadException(shaderSource);
        }

		LOGGER.finer(resultingShaderSource);
        GpuContext.exitOnGLError("loadShader");

        return shader;
    }

    void setShaderSource(CodeSource shaderSource);
    CodeSource getShaderSource();

    static TypedTuple<String, Integer> replaceIncludes(String shaderFileAsText, int currentNewLineCount) throws IOException {

        Pattern includePattern = Pattern.compile("//include\\((.*)\\)");
        Matcher includeMatcher = includePattern.matcher(shaderFileAsText);

        while (includeMatcher.find()) {
            String filename = includeMatcher.group(1);
            String fileToInclude = FileUtils.readFileToString(new File(getDirectory() + filename));
            currentNewLineCount += Util.countNewLines(fileToInclude);
            shaderFileAsText = shaderFileAsText.replaceAll(String.format("//include\\(%s\\)", filename),
                    fileToInclude);
        }

        return new TypedTuple<>(shaderFileAsText, new Integer(currentNewLineCount));
    }

    static String replaceLineNumbersWithDynamicLinesAdded(String shaderInfoLog, int newlineCount) {

		Pattern loCPattern = Pattern.compile("\\((\\w+)\\) :");
		Matcher loCMatcher = loCPattern.matcher(shaderInfoLog);

		while (loCMatcher.find()) {
			String oldLineNumber = loCMatcher.group(1);
			int newLineNumber = Integer.parseInt(oldLineNumber) - newlineCount;
			shaderInfoLog = shaderInfoLog.replaceAll(String.format("\\(%s\\) :", oldLineNumber), String.format("(%d) :", newLineNumber));
		}

		return shaderInfoLog;
	}

    static String getDirectory() {
		return DirectoryManager.WORKDIR_NAME + "/assets/shaders/";
	}

    void setId(int shaderID);
    int getId();

    OpenGLShader getShaderType();

    enum OpenGLShader {
        VertexShader(GL20.GL_VERTEX_SHADER),
        FragmentShader(GL20.GL_FRAGMENT_SHADER),
        GeometryShader(GL32.GL_GEOMETRY_SHADER),
        ComputeShader(GL43.GL_COMPUTE_SHADER);

        public final int glShaderType;

        OpenGLShader(int glShaderType) {
            this.glShaderType = glShaderType;
        }
    }

    final class ShaderSourceFactory {
        public static CodeSource getShaderSource(String shaderSource) {
            if("".equals(shaderSource)) {
                return null;
            }
            return new CodeSource(shaderSource);
        }

        public static CodeSource getShaderSource(File file) {
            if(file.exists()) {
                try {
                    return new CodeSource(file);
                } catch (IOException e) {
                    e.printStackTrace();
                    return null;
                }
            }

            LOGGER.warning("File doesn't exist: " + file.getAbsolutePath());
            return null;
        }
    }

    class ShaderLoadException extends RuntimeException {
        private final CodeSource shaderSource;

        public ShaderLoadException(CodeSource shaderSource) {
            this.shaderSource = shaderSource;
        }

        public String toString() {
            return shaderSource.getFilename() + "\n" + shaderSource.getSource();
        }

    }
}

File: engine/src/main/java/de/hanno/hpengine/engine/graphics/shader/Shader.kt
package de.hanno.hpengine.engine.graphics.shader

import de.hanno.hpengine.engine.directory.EngineDirectory
import de.hanno.hpengine.engine.graphics.shader.define.Defines
import de.hanno.hpengine.util.TypedTuple
import de.hanno.hpengine.util.Util
import de.hanno.hpengine.util.ressources.CodeSource
import de.hanno.hpengine.util.ressources.FileBasedCodeSource
import de.hanno.hpengine.util.ressources.Reloadable
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL20
import java.io.IOException
import java.util.logging.Logger
import java.util.regex.Pattern

sealed class Shader(private val programManager: ProgramManager<*>,
                    var source: CodeSource,
                    val defines: Defines = Defines(),
                    val shaderType: ShaderType) : Reloadable {
    private val gpuContext = programManager.gpuContext

    val id = programManager.gpuContext { GL20.glCreateShader(shaderType.glShaderType) } // TODO: Abstract createShader in gpuContext

    override fun load() {
        source.load()

        val resultingShaderSource = programManager.run { source.toResultingShaderSource(defines) }

        gpuContext.invoke {
            GL20.glShaderSource(id, resultingShaderSource)
            GL20.glCompileShader(id)
        }

        val shaderLoadFailed = gpuContext.invoke {
            val shaderStatus = GL20.glGetShaderi(id, GL20.GL_COMPILE_STATUS)
            if (shaderStatus == GL11.GL_FALSE) {
                System.err.println("Could not compile " + shaderType + ": " + source.name)
                var shaderInfoLog = GL20.glGetShaderInfoLog(id, 10000)
                shaderInfoLog = replaceLineNumbersWithDynamicLinesAdded(shaderInfoLog, resultingShaderSource.lines().size)
                ShaderLoadException(shaderInfoLog)
            } else null
        }

        shaderLoadFailed?.let { throw it }
    }

    override fun reload() = load()
    override fun unload() {
        source.unload()
        GL20.glDeleteShader(id)
    }

    override val name: String
        get() = source.name

    override fun equals(other: Any?): Boolean {
        if(other !is Shader) return false

        return id == other.id && defines == other.defines
    }

    override fun hashCode(): Int {
        var hash = 0
        hash += id.hashCode()
        hash += defines.hashCode()
        return hash
    }
    init {
        load()
    }

    companion object {

        val LOGGER = Logger.getLogger(Shader::class.java.name)

        @Throws(IOException::class)
        fun replaceIncludes(engineDir: EngineDirectory, shaderFileAsText: String, currentNewLineCount: Int): TypedTuple<String, Int> {
            var shaderFileAsText = shaderFileAsText
            var currentNewLineCount = currentNewLineCount

            val includePattern = Pattern.compile("//include\\((.*)\\)")
            val includeMatcher = includePattern.matcher(shaderFileAsText)

            while (includeMatcher.find()) {
                val filename = includeMatcher.group(1)
                val fileToInclude = engineDir.resolve("shaders/$filename").readText()
                currentNewLineCount += Util.countNewLines(fileToInclude)
                shaderFileAsText = shaderFileAsText.replace(String.format("//include\\(%s\\)", filename).toRegex(), fileToInclude)
            }

            return TypedTuple(shaderFileAsText, currentNewLineCount)
        }

        fun replaceLineNumbersWithDynamicLinesAdded(shaderInfoLog: String, newlineCount: Int): String {
            var shaderInfoLog = shaderInfoLog

            val loCPattern = Pattern.compile("\\((\\w+)\\) :")
            val loCMatcher = loCPattern.matcher(shaderInfoLog)

            while (loCMatcher.find()) {
                val oldLineNumber = loCMatcher.group(1)
                val newLineNumber = Integer.parseInt(oldLineNumber) - newlineCount
                val regex = String.format("\\($oldLineNumber\\) :", oldLineNumber).toRegex()
                shaderInfoLog = shaderInfoLog.replace(regex, "(ln $newLineNumber) :")
            }

            return shaderInfoLog
        }

    }
}

class VertexShader(programManager: OpenGlProgramManager, shaderSource: CodeSource, defines: Defines = Defines()) : Shader(programManager, shaderSource, defines, ShaderType.VertexShader)
class TesselationControlShader(programManager: OpenGlProgramManager, shaderSource: CodeSource, defines: Defines = Defines()) : Shader(programManager, shaderSource, defines, ShaderType.TesselationControlShader)
class TesselationEvaluationShader(programManager: OpenGlProgramManager, shaderSource: CodeSource, defines: Defines = Defines()) : Shader(programManager, shaderSource, defines, ShaderType.TesselationEvaluationShader)
class GeometryShader(programManager: OpenGlProgramManager, shaderSource: CodeSource, defines: Defines = Defines()) : Shader(programManager, shaderSource, defines, ShaderType.GeometryShader)
class FragmentShader(programManager: OpenGlProgramManager, shaderSource: CodeSource, defines: Defines = Defines()) : Shader(programManager, shaderSource, defines, ShaderType.FragmentShader)

class ComputeShader(programManager: OpenGlProgramManager, shaderSource: CodeSource, defines: Defines = Defines()) : Shader(programManager, shaderSource, defines, ShaderType.ComputeShader)
